wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointPlugin.java

```diff
@@ -0,0 +1,370 @@
+import software.amazon.jdbc.util.WrapperUtils;
+import software.amazon.jdbc.util.telemetry.TelemetryCounter;
+import software.amazon.jdbc.util.telemetry.TelemetryFactory;
+
+/**
+ * A plugin that analyzes custom endpoints for custom endpoint information and custom endpoint changes, such as adding
+ * or removing an instance in the custom endpoint.
+ */
+public class CustomEndpointPlugin extends AbstractConnectionPlugin {
+  private static final Logger LOGGER = Logger.getLogger(CustomEndpointPlugin.class.getName());
+  private static final String TELEMETRY_CUSTOM_ENDPOINT_COUNTER = "customEndpoint.endpointConnections.counter";
+  private static final String TELEMETRY_WAIT_FOR_INFO_COUNTER = "customEndpoint.waitForInfo.counter";
+
+  protected static final long CACHE_CLEANUP_RATE_NANO = TimeUnit.MINUTES.toNanos(1);
+  protected static final SlidingExpirationCacheWithCleanupThread<String, CustomEndpointMonitor> monitors =
+      new SlidingExpirationCacheWithCleanupThread<>(
+          CustomEndpointMonitor::shouldDispose,
+          (monitor) -> {
+            try {
+              monitor.close();
+            } catch (Exception ex) {
+              // ignore
+            }
+          },
+          CACHE_CLEANUP_RATE_NANO);
+
+  private static final Set<String> subscribedMethods =
+      Collections.unmodifiableSet(new HashSet<String>() {
+        {
+          addAll(SubscribedMethodHelper.NETWORK_BOUND_METHODS);
+          add("connect");
+          add("forceConnect");
+        }
+      });
+
+  public static final AwsWrapperProperty CUSTOM_ENDPOINT_INFO_REFRESH_RATE_MS = new AwsWrapperProperty(
+      "customEndpointInfoRefreshRateMs", "30000",
+      "Controls how frequently custom endpoint monitors fetch custom endpoint info.");
+
+  public static final AwsWrapperProperty WAIT_FOR_CUSTOM_ENDPOINT_INFO_TIMEOUT_MS = new AwsWrapperProperty(
+      "waitForCustomEndpointInfoTimeoutMs", "5000",
+      "Controls the maximum amount of time that the plugin will wait for custom endpoint info to be made "
+          + "available by the custom endpoint monitor.");
+
+  public static final AwsWrapperProperty CUSTOM_ENDPOINT_MONITOR_IDLE_EXPIRATION_MS = new AwsWrapperProperty(
+      "customEndpointMonitorExpirationMs", String.valueOf(TimeUnit.MINUTES.toMillis(15)),
+      "Controls how long a monitor should run without use before expiring and being removed.");
+
+  public static final AwsWrapperProperty REGION_PROPERTY = new AwsWrapperProperty(
+      "customEndpointRegion", null,
+      "The region of the cluster's custom endpoints.");
+
+  static {
+    PropertyDefinition.registerPluginProperties(CustomEndpointPlugin.class);
+  }
+
+  protected final PluginService pluginService;
+  protected final Properties props;
+  protected final RdsUtils rdsUtils = new RdsUtils();
+  protected final BiFunction<HostSpec, Region, RdsClient> rdsClientFunc;
+
+  private final TelemetryCounter endpointConnectionsCounter;
```

Comment:
Why is it private and not protected?

Reply:
That was just from a copy-paste, I'll change to protected